src/build.rs

```diff
@@ -998,6 +997,54 @@ fn deprecation_warnings(ws: &Workspace, args: &ArgMatches) -> anyhow::Result<()>
     Ok(())
 }
 
+/// What library types to build
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct LibraryTypes {
+    pub(crate) staticlib: bool,
+    pub(crate) cdylib: bool,
+}
+
+impl LibraryTypes {
+    fn from_target(target: &target::Target) -> Self {
+        Self {
+            staticlib: true,
+            cdylib: target.os != "none" && target.env != "musl",
+        }
+    }
+
+    fn from_config(target: &target::Target, args: &ArgMatches) -> Self {
+        match args.get_many::<String>("library-type") {
+            Some(library_types) => Self::from_library_types(target, library_types),
+            None => Self::from_target(target),
+        }
+    }
+
+    pub(crate) fn from_library_types<S: AsRef<str>>(
+        target: &target::Target,
+        library_types: impl Iterator<Item = S>,
+    ) -> Self {
+        let (mut staticlib, mut cdylib) = (false, false);
+
+        for library_type in library_types {
+            staticlib |= library_type.as_ref() == "staticlib";
+            cdylib |= library_type.as_ref() == "cdylib";
+        }
+
+        // in this case, a cdylib cannot be produced
+        cdylib &= target.os != "none";
```

Comment:
based on the original logic, there is no check here for "musl", but there is in `from_target`. Under what circumstances can musl build a cdylib?

Reply:
when musl is being used as main libc, e.g. in Alpine, Gentoo and OpenWRT :)

Reply:
cool, I added some comments with this information